pkg/log/logger.go

```diff
@@ -0,0 +1,262 @@
+			outputFile := filepath.Join(v.config.KubernetesDir, sandboxId, filename)
+
+			if _, err := os.Stat(filepath.Dir(outputFile)); os.IsNotExist(err) {
+				glog.V(1).Infof("Kubernetes directory '%s' does not exist: %s", filepath.Dir(outputFile), err.Error())
+				break
+			}
+
+			go v.workerRunner.RunNewWorker(inputFile, outputFile, fmt.Sprintf("%s/%s", sandboxId, filename))
+		}
+	}
+
+	return nil
+}
+
+func (v *virtletLogger) StopObsoleteWorkers() {
+	glog.V(1).Infoln("Stop obsolete workers")
+	for _, workerId := range v.workerRunner.ListWorkers() {
+		if _, err := os.Stat(filepath.Join(v.config.KubernetesDir, workerId)); os.IsNotExist(err) {
+			v.workerRunner.StopWorker(workerId)
+			continue
+		}
+		if _, err := os.Stat(filepath.Join(v.config.VirtletDir, workerId)); os.IsNotExist(err) {
+			v.workerRunner.StopWorker(workerId)
+		}
+	}
+}
+
+func (v *virtletLogger) StopAllWorkers() {
+	v.workerRunner.StopAllWorkers()
+}
+
+const STOP string = "---LOGGER-STOP---"
+
+type workerRunner struct {
+	workers    map[string]chan *tail.Line // map[sandboxId]chan
+	workersMux sync.Mutex
+	statusCh   chan string
+}
+
+func newWorkerRunner() *workerRunner {
+	return &workerRunner{
+		workers:  make(map[string]chan *tail.Line),
+		statusCh: nil,
+	}
+}
+
+func (w *workerRunner) RunNewWorker(inputFile, outputFile, workerId string) {
+	glog.V(1).Infoln("Spawned worker", workerId)
+
+	w.reportWorkerState("START", workerId)
+	defer w.reportWorkerState("STOP", workerId)
+
+	if _, err := os.Stat(outputFile); os.IsNotExist(err) {
+		file, err := os.Create(outputFile)
+		if err != nil {
+			glog.V(1).Infoln("Failed to create output file:", err)
+			return
+		}
+		file.Close()
+	}
+
+	f, err := os.OpenFile(outputFile, os.O_WRONLY, 0777)
+	if err != nil {
+		glog.V(1).Infoln("Failed to open output file:", err)
+		return
+	}
+	defer f.Close()
+
+	// Tail VM's file.
+	t, err := tail.TailFile(inputFile, tail.Config{Follow: true, ReOpen: true})
+	if err != nil {
+		glog.V(1).Infoln("Failed to tail input file:", err)
+		return
+	}
+
+	// Expose channel to virtletLogger so that it can close it when needed.
+	w.registerWorker(workerId, t.Lines)
+
+	// Do work. This forloop will block until canceled; it will wait for new
+	// lines to come and parse them immediately.
+	for line := range t.Lines {
+		if line.Text == STOP {
+			break
+		}
+
+		if line.Err != nil {
+			glog.V(1).Infoln("Error reading line:", line.Err.Error())
+			break
+		}
+
+		// Convert raw line into Kubernetes json.
+		converted := fmt.Sprintf(`{"time": "%s", "stream": "stdout","log":"%s\n"}`, line.Time.Format(time.RFC3339), escapeLine(line.Text))
+		converted = converted + "\n"
+
+		f.WriteString(converted)
+		f.Sync()
+
+		w.reportWorkerState("RUN", workerId)
+	}
+
+	// This code is only executed when t.Lines channel is closed.
+	delete(w.workers, workerId)
+	glog.V(1).Infof("Worker for sandbox '%s' stopped gracefully\n", workerId)
```

Comment:
Done for all occurances of `glog.V(1).Infof()`. Sorry for not testing this after replacing `fmt` with `glog`!